modules/reindex/src/test/java/org/elasticsearch/reindex/ReindexPluginMetricsTests.java

```diff
@@ -0,0 +1,131 @@
+    }
+
+    protected TestTelemetryPlugin getTestTelemetryPlugin() {
+        return (internalCluster().getInstance(PluginsService.class).filterPlugins(TestTelemetryPlugin.class).toList().get(0));
+    }
+
+    public void testFiltering() throws Exception {
+        internalCluster().startNode();
+
+        indexRandom(
+            true,
+            prepareIndex("source").setId("1").setSource("foo", "a"),
+            prepareIndex("source").setId("2").setSource("foo", "a"),
+            prepareIndex("source").setId("3").setSource("foo", "b"),
+            prepareIndex("source").setId("4").setSource("foo", "c")
+        );
+        assertHitCount(prepareSearch("source").setSize(0), 4);
+
+        final String dataNodeName = internalCluster().getRandomNodeName();
+        final TestTelemetryPlugin testTelemetryPlugin = internalCluster().getInstance(PluginsService.class, dataNodeName)
+            .filterPlugins(TestTelemetryPlugin.class)
+            .findFirst()
+            .orElseThrow();
+
+        // TestTelemetryPlugin testTelemetryPlugin = getTestTelemetryPlugin();
+        // testTelemetryPlugin.resetMeter();
+
+        // Copy all the docs
+        ReindexRequestBuilder copy = reindex().source("source").destination("dest").refresh(true);
+        assertThat(copy.get(), matcher().created(4));
+
+        final var tp = internalCluster().getInstance(ThreadPool.class, dataNodeName);
+        final var tps = new ThreadPoolStats[1];
+
+        // wait for all threads to complete so that we get deterministic results
+        waitUntil(() -> (tps[0] = tp.stats()).stats().stream().allMatch(s -> s.active() == 0));
+
+        testTelemetryPlugin.collect();
+        assertHitCount(prepareSearch("dest").setSize(0), 4);
+        List<Measurement> measurements = testTelemetryPlugin.getLongHistogramMeasurement(REINDEX_TIME_HISTOGRAM);
+        assertThat(measurements.size(), equalTo(1));
+
+        // Now none of them
+        createIndex("none");
+        copy = reindex().source("source").destination("none").filter(termQuery("foo", "no_match")).refresh(true);
+        assertThat(copy.get(), matcher().created(0));
+
+        // wait for all threads to complete so that we get deterministic results
+        waitUntil(() -> (tps[0] = tp.stats()).stats().stream().allMatch(s -> s.active() == 0));
+
+        testTelemetryPlugin.collect();
+        assertHitCount(prepareSearch("none").setSize(0), 0);
+        measurements = testTelemetryPlugin.getLongHistogramMeasurement(REINDEX_TIME_HISTOGRAM);
+        assertThat(measurements.size(), equalTo(2));
+
+        // Now half of them
+        copy = reindex().source("source").destination("dest_half").filter(termQuery("foo", "a")).refresh(true);
+        assertThat(copy.get(), matcher().created(2));
+
+        // wait for all threads to complete so that we get deterministic results
+        waitUntil(() -> (tps[0] = tp.stats()).stats().stream().allMatch(s -> s.active() == 0));
```

Comment:
usually what we do when an assertion might need to wait until something completes in the background is use `assertBusy` which asserts a few times until either the assertion is true or fails once the timeout is reached.

Reply:
Ah ok, so I don't need to do the explicit wait if I use assertBusy () ?

Reply:
No, you don't need to do the explicit wait, `assertBusy` waits implicitly.